sqlglot/optimizer/qualify_columns.py

```diff
@@ -0,0 +1,321 @@
+    for node, parent, _ in expression.walk(stop_types=exp.Select):
+        if node is expression:
+            continue  # Skip this node itself - we only care about children
+        if isinstance(node, exp.CTE):
+            ctes.append(node)
+        elif isinstance(node, (exp.Select, exp.Union)):
+            if isinstance(parent, exp.Subquery):
+                derived_tables.append(parent)
+            else:
+                subqueries.append(node)
+        elif isinstance(node, exp.Table):
+            tables.append(node)
+        elif isinstance(node, exp.Column):
+            columns.append(node)
+
+    parent_selectables = _merge(
+        parent_selectables,
+        _qualify_derived_tables(ctes, schema, sequence, parent_selectables, chain=True),
+    )
+    selectables = _merge({}, _qualify_tables(tables, schema, parent_selectables))
+    selectables = _merge(
+        selectables,
+        _qualify_derived_tables(derived_tables, schema, sequence, parent_selectables),
+    )
+    expansions = _expand_stars(select_stars, selectables)
+    columns.extend(expansions)
+    selections.extend(expansions)
+    selectables = _merge(parent_selectables, selectables)
+    _qualify_columns(columns, selectables)
+    _qualify_subqueries(subqueries, schema, sequence, selectables)
+    return _qualify_outputs(selections)
+
+
+def _qualify_derived_tables(
+    derived_tables, schema, sequence, parent_selectables, chain=False
+):
+    selectables = {}
+    for subquery in derived_tables:
+        subquery_outputs = _qualify_union(
+            expression=subquery.this,
+            schema=schema,
+            sequence=sequence,
+            parent_selectables=_merge(parent_selectables, selectables)
+            if chain
+            else parent_selectables,
+        )
+
+        table_alias = subquery.args.get("alias")
+        if not table_alias:
+            table_alias = exp.TableAlias()
+            subquery.set("alias", table_alias)
+
+        alias = table_alias.args.get("this")
+        if not alias:
+            alias = exp.to_identifier(f"_alias{sequence.next()}")
+            table_alias.set("this", alias)
+
+        alias_columns = table_alias.args.get("columns")
+        if not alias_columns:
+            alias_columns = [exp.to_identifier(o) for o in subquery_outputs]
+            table_alias.set("columns", alias_columns)
+
+        selectable_table = alias.text("this")
+
+        selectable_columns = [c.text("this") for c in alias_columns]
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+
+    return selectables
+
+
+def _qualify_tables(tables, schema, parent_selectables):
+    selectables = {}
+    for table in tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
```

Comment:
so you add alias to select x AS x even if x == x, to be consistent, we should probably do FROM x AS x as well -- or if we want a cleaner canonical form, we strip out aliases if alias == source. thougths?

Reply:
I think there's a balance between _pretty_ and _consistent_.

In my mind, canonical SQL should prioritize consistency over prettiness. So I thought it was reasonable to coerce all the selections to the Alias type.

Reply:
sounds good